src/libraries/System.Private.CoreLib/src/System/Number.Parsing.cs

```diff
@@ -1388,7 +1389,27 @@ internal enum ParsingStatus
         internal static void ThrowFormatException<TChar>(ReadOnlySpan<TChar> value)
             where TChar : unmanaged, IUtfChar<TChar>
         {
-            throw new FormatException(SR.Format(SR.Format_InvalidStringWithValue, value.ToString()));
+            string errorMessage;
+
+            if (typeof(TChar) == typeof(byte))
+            {
+                // Decode the UTF8 value into a string we can include in the error message. We're here
+                // because we failed to parse, which also means the bytes might not be valid UTF8,
+                // so fallback to a message that doesn't include the value if the bytes are invalid.
+                // It's possible after we check the bytes for validity that they could be concurrently
+                // mutated, but if that's happening, all bets are off, anyway, and it simply impacts
+                // which exception is thrown.
+                ReadOnlySpan<byte> bytes = MemoryMarshal.Cast<TChar, byte>(value);
+                errorMessage = Utf8.IsValid(bytes) ?
+                    SR.Format(SR.Format_InvalidStringWithValue, Encoding.UTF8.GetString(bytes)) :
+                    SR.Format_InvalidString;
```

Comment:
I know this is an exception path and therefore likely cold, but is it worth just calling `Utf8.ToUtf16` instead and checking the `OperationStatus`, since its cheaper than using `Utf8.IsValid` and then trying to extract the string?

Reply:
> but is it worth just calling Utf8.ToUtf16 instead and checking the OperationStatus

I'd need a sufficiently large destination buffer; didn't seem worth doing the whole stackalloc/ArrayPool dance to get one. Is there something else you had in mind?

Reply:
> We're getting that allocation anyways with Encoding.UTF8.GetString

We shouldn't be. If the input is <= 32 bytes, it'll stackalloc, and if it's longer, it'll use GetCharCount to get the actual length, allocate the string of that exact length, and then decode directly into the string.

> The allocation is probably cheaper than the IsValid check and doesn't require the dance.

| Method     | Length | Mean     | Allocated |
|----------- |------- |---------:|----------:|
| GetString1 | 5      | 23.64 ns |      32 B |
| GetString2 | 5      | 22.61 ns |      72 B |
| GetString1 | 100    | 47.10 ns |     224 B |
| GetString2 | 100    | 59.77 ns |     456 B |

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Text;
using System.Text.Unicode;

BenchmarkSwitcher.FromAssembly(typeof(Tests).Assembly).Run(args);

[HideColumns("Error", "StdDev", "Median", "RatioSD")]
[MemoryDiagnoser(false)]
public partial class Tests
{
    private byte[] _data;

    [Params(5, 100)]
    public int Length { get; set; }

    [GlobalSetup]
    public void Setup() => _data = Encoding.UTF8.GetBytes(new string('a', Length));

    [Benchmark]
    public string GetString1()
    {
        if (Utf8.IsValid(_data))
        {
            return Encoding.UTF8.GetString(_data);
        }
        return "";
    }

    [Benchmark]
    public string GetString2()
    {
        char[] dest = new char[Encoding.UTF8.GetMaxCharCount(_data.Length)];
        if (Utf8.ToUtf16(_data, dest, out _, out int charsWritten) == System.Buffers.OperationStatus.Done)
        {
            return dest.AsSpan(0, charsWritten).ToString();
        }
        return "";
    }
}
```